internal/checker/symbolaccessibility.go

```diff
@@ -0,0 +1,762 @@
+		return candidates[0]
+	}
+	return nil
+}
+
+func (ch *Checker) getAccessibleSymbolChain(
+	symbol *ast.Symbol,
+	enclosingDeclaration *ast.Node,
+	meaning ast.SymbolFlags,
+	useOnlyExternalAliasing bool,
+) []*ast.Symbol {
+	return ch.getAccessibleSymbolChainEx(accessibleSymbolChainContext{symbol, enclosingDeclaration, meaning, useOnlyExternalAliasing, make(map[ast.SymbolId]map[uintptr]struct{})})
+}
+
+func (ch *Checker) GetAccessibleSymbolChain(
+	symbol *ast.Symbol,
+	enclosingDeclaration *ast.Node,
+	meaning ast.SymbolFlags,
+	useOnlyExternalAliasing bool,
+) []*ast.Symbol {
+	return ch.getAccessibleSymbolChain(symbol, enclosingDeclaration, meaning, useOnlyExternalAliasing)
+}
+
+type accessibleSymbolChainContext struct {
+	symbol                  *ast.Symbol
+	enclosingDeclaration    *ast.Node
+	meaning                 ast.SymbolFlags
+	useOnlyExternalAliasing bool
+	visitedSymbolTablesMap  map[ast.SymbolId]map[uintptr]struct{}
+}
+
+func (ch *Checker) getAccessibleSymbolChainEx(ctx accessibleSymbolChainContext) []*ast.Symbol {
+	if ctx.symbol == nil {
+		return nil
+	}
+	if isPropertyOrMethodDeclarationSymbol(ctx.symbol) {
+		return nil
+	}
+	// Go from enclosingDeclaration to the first scope we check, so the cache is keyed off the scope and thus shared more
+	var firstRelevantLocation *ast.Node
+	ch.someSymbolTableInScope(ctx.enclosingDeclaration, func(_ ast.SymbolTable, _ bool, _ bool, node *ast.Node) bool {
+		firstRelevantLocation = node
+		return true
+	})
+	links := ch.symbolContainerLinks.Get(ctx.symbol)
+	linkKey := accessibleChainCacheKey{ctx.useOnlyExternalAliasing, firstRelevantLocation, ctx.meaning}
+	if links.accessibleChainCache == nil {
+		links.accessibleChainCache = make(map[accessibleChainCacheKey][]*ast.Symbol)
+	}
+	existing, ok := links.accessibleChainCache[linkKey]
+	if ok {
+		return existing
+	}
+
+	var result []*ast.Symbol
+
+	ch.someSymbolTableInScope(ctx.enclosingDeclaration, func(t ast.SymbolTable, ignoreQualification bool, isLocalNameLookup bool, _ *ast.Node) bool {
+		res := ch.getAccessibleSymbolChainFromSymbolTable(ctx, t, ignoreQualification, isLocalNameLookup)
+		if len(res) > 0 {
+			result = res
+			return true
+		}
+		return false
+	})
+	links.accessibleChainCache[linkKey] = result
+	return result
+}
+
+/**
+* @param {ignoreQualification} boolean Set when a symbol is being looked for through the exports of another symbol (meaning we have a route to qualify it already)
+ */
+func (ch *Checker) getAccessibleSymbolChainFromSymbolTable(ctx accessibleSymbolChainContext, t ast.SymbolTable, ignoreQualification bool, isLocalNameLookup bool) []*ast.Symbol {
+	symId := ast.GetSymbolId(ctx.symbol)
+	visitedSymbolTables, ok := ctx.visitedSymbolTablesMap[symId]
+	if !ok {
+		visitedSymbolTables = make(map[uintptr]struct{})
+		ctx.visitedSymbolTablesMap[symId] = visitedSymbolTables
+	}
+
+	id := reflect.ValueOf(t).Pointer() // TODO: Is this seriously the only way to check reference equality of maps?
```

Comment:
This is definitely unsafe. The symbol tables could get GC'd because this is hiding the pointer behind a `uintptr`. Do we really need to be checking reference equality?

The only way that this could be okay is if we used `.UnsafePointer()` (the recommended function) instead of `uintptr` to ensure we actually have a reference. But, this is all quite spooky.

Reply:
We use the identity of the symbol tables to check which ones we've already traversed, and by rights they should never mutate _or_ be GC'd under us here, since we should have calculated and cached the tables by this point (which should only lose their last referrer when the checker itself is collected). Can't use symbols or symbol ids for identity, since they contain multiple tables. And exhaustive map comparison is obviously a nonstarter, as these are often huge. I can swap the pointer kind in use, if that'll help, though.